pkg/agent/tcpconn/queue.go

```diff
@@ -0,0 +1,172 @@
+package tcpconn
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"math/rand"
+
+	"github.com/florianl/go-nfqueue"
+	"github.com/grafana/xk6-disruptor/pkg/iptables"
+	"github.com/grafana/xk6-disruptor/pkg/runtime"
+)
+
+// NFQConfig contains netfilter queue IDs that iptables rules and the packet handler use to communicate.
+type NFQConfig struct {
+	// QueueID is an arbitrary integer used to identify a queue where a handler listens and a disruptor redirects target
+	// packets.
+	QueueID uint16
+	// RejectMark is an arbitrary integer which the handler uses to mark packets that need to be dropped.
+	RejectMark uint32
+}
+
+// RandomNFQConfig returns a NFQConfig with two random integers to be used as queue IDs and reject mark.
+// To ensure the numbers are not zero they are ORed with 0b1, as adding 1 can actually result in the number overflowing
+// and becoming zero.
+func RandomNFQConfig() NFQConfig {
+	return NFQConfig{
+		QueueID:    uint16(rand.Int31()) | 0b1,
+		RejectMark: uint32(rand.Int31()) | 0b1,
+	}
+}
+
+// Queue is the interface implemented by objects that can write packets to a channel.
+type Queue interface {
+	// Start starts listening to packets, blocking until an error occurs or the context is cancelled.
+	Start(ctx context.Context, packets chan<- Packet) error
+}
+
+// Packet is a packet received by a Queue. It can describe itself and be accepted or rejected.
+type Packet interface {
+	// Bytes returns the raw bytes that make the packet.
+	Bytes() []byte
+	// Accept marks the packet to be accepted.
+	Accept()
+	// Reject marks the packet to be rejected.
+	Reject()
+}
+
+// FakeQueue is a fake implementation of queue. It stalls until the context is cancelled, never sending any packet.
+type FakeQueue struct{}
+
+// Start implements the Queue interface.
+func (f FakeQueue) Start(ctx context.Context, _ chan<- Packet) error {
+	<-ctx.Done()
+	return ctx.Err()
+}
+
+// NFQueue implements the Queue interface using netfilter's nfqueue mechanism, reading packets sent to userspace by
+// netfilter. NFQueue will process packets sent to the NFQUEUE chain with queue id specified in NFQConfig.QueueID.
+// Accept()ed packets are immediately forwarded to the ACCEPT chain and do not traverse subsequent iptables rules.
+// Reject()ed packets are requeued with the mark specified in NFQConfig.RejectMark.
+// Iptables rules should take care to _not_ send again to the NFQUEUE target packets that have been marked with the
+// RejectMark, and should direct those to the REJECT chain instead.
+type NFQueue struct {
+	Executor   runtime.Executor
```

Comment:
Good point.

Reply:
Good point. The reason why I didn't do it this way is because I didn't want to give the impression that the `Iptables` object is a reusable wrapper around iptables: It is not, it is intended to be completely owned by whoever needs to add and then remove iptables rules. It would be very bad if two objects attempted to use the same `Iptables` instance at the same time (either concurrently, or serially but on different parts of the code).

We can document this instead, or perhaps make an intermediate object (could be named a `RuleSet`) that better conveys this intention (as I'm writing this second option I'm liking it quite a lot). WDYT?

Reply:
>  It would be very bad if two objects attempted to use the same Iptables instance at the same time (either concurrently, or serially but on different parts of the code).

This is an important limitation. I haven't considered it.

> We can document this instead, or perhaps make an intermediate object (could be named a RuleSet) that better conveys this intention (as I'm writing this second option I'm liking it quite a lot). WDYT?

I like the idea of the ruleset and make the Iptables wrapper stateless (store the rules in the ruleset)
But this is a significant change, maybe we can move to a follow-up PR.

Reply:
Makes sense, we can add the ruleset change and implement it in the redirector in the same PR.